Prepare the introduction of more keeper faults

Algunenano opened the pull request:
### Changelog category (leave one):
- Not for changelog

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
- Prepares the introduction of more keeper faults


Longer details:
* Refactors ZooKeeperWithFaultInjection to prepare to introduce it in more places:
  * Implementation goes to the .cpp file.
  * Adds a bunch of new methods with faults to be used in replicated merge tree
  * Temporarily maintains the tracking of ephemeral nodes created with a faulty keeper. In the future, once "everything" handles retries this should removed in favor of force a session reload, which is what happens in a real environment.
  * Introduces faults in async functions too via promises.

* It reworks how part_committed_locally_but_zookeeper (old) / resolve_uncertain_commit_stage (less old) worked so that we can keep retrying in case the ZK session failed when committing. Before the uncertain stage could only be resolved if ZK failed after write (and in that case everything was good) but it wouldn't recover it data wasn't written to keeper. Now it recovers both cases.

* Fixes an issue found related with ephemeral nodes cleaning up in backups fault and changes the name (`handleEphemeralNodeExistence` ->  `deleteEphemeralNodeIfContentMatches`) to avoid further confusions. I included in this PR since changing the name and introducing faults is what leads to noticing the problem and enables fixing them only once (and not twice, one in a different PR and another time in the refactor)

AFAICS. Closes https://github.com/ClickHouse/ClickHouse/issues/50465 as it removed the special handling of tryMulti() responses

cc @devcrafter

Algunenano committed:
Prepare the introduction of more keeper faults

Algunenano committed:
WIP: Move implementation

Algunenano committed:
Refactor ZooKeeperWithFaultInjection

Algunenano committed:
Style

Algunenano committed:
Style and fix log level

Algunenano committed:
Merge remote-tracking branch 'blessed/master' into backup_1

Algunenano commented:
It seems I will need to bring more commits from https://github.com/ClickHouse/ClickHouse/pull/56389. As the tests run in parallel, the moment you enable more faults (some ops had some disabled, and the refactor got rid of that) they all clash with each other more frequently so you see the lack of proper recovery more often, e.g. the `part_committed_locally_but_zookeeper` thingy.

Algunenano committed:
Merge remote-tracking branch 'blessed/master' into backup_1

Algunenano committed:
CI: Increase keeper retries and decrease retry backoff to a minimum

Algunenano committed:
Replace part_committed_locally_but_zookeeper with retries

Algunenano committed:
Recover special handling of ephemeral nodes in ZooKeeperWithFaultInjection

Algunenano commented:
* Fixed broken debug build. I'm not sure why it compiled in other builds when I used `ch_assert` vs `chassert`
* The commit "fixing" 00509_extended_storage_definition_syntax_zookeeper wasn't necessary as, in the end, it wasn't unrelated. OPTIMIZE failed because there were inserts on the fly because of dangling ephemeral nodes due to keeper faults. After fixing that it does not fail

Algunenano committed:
Tidy

Algunenano committed:
Fix WithRetries callback

Algunenano committed:
Fix bug in exists()

Algunenano commented:
Errors:
* `AST fuzzer (asan)` -> Related. Bug in ZooKeeperWithFaultInjection::exists logger.
* `Integration tests (asan) [1/4]` -> PG leak -> https://github.com/ClickHouse/ClickHouse/issues/56627
* `Integration tests (release) [2/4] ` -> test_backup_restore_on_cluster/test_concurrency.py::test_create_or_drop_tables_during_backup[Replicated-ReplicatedMergeTree]. Looks related to retries or extra faults added, having a look
* `Stateless tests (debug) [2/5]` -> `01280_ttl_where_group_by`. Flaky
* `Stateless tests (debug, s3 storage) [3/6]` ->  `01280_ttl_where_group_by` again
* `Stateless tests (release, analyzer)`. `01287_max_execution_speed` and `01600_parts_states_metrics_long` which are flaky with the analyzer.
* `Stateless tests (tsan, s3 storage) [2/5] ` ->  `01280_ttl_where_group_by` again

Algunenano committed:
Merge remote-tracking branch 'blessed/master' into backup_1

Algunenano commented:
Failures:
* `Stateless tests (debug, s3 storage) [3/6]` -> 01280_ttl_where_group_by. Unrelated
* `Stress test (debug)` -> https://github.com/ClickHouse/ClickHouse/issues/57127. Unrelated

Algunenano commented on src/Storages/MergeTree/ReplicatedMergeTreeSink.cpp:
```diff
@@ -952,32 +923,47 @@ std::pair<std::vector<String>, bool> ReplicatedMergeTreeSinkImpl<async_insert>::
-                {
-                    my_storage.enqueuePartForCheck(part_name, MAX_AGE_OF_LOCAL_PART_THAT_WASNT_ADDED_TO_ZOOKEEPER);
-                });
+            LOG_TRACE(
```
@tavplubix Do you mind having a look at this changes, specially checking that `rename_part_to_temporary();` in the case where the data wasn't written to keeper (to retry) is safe? Thanks!

devcrafter commented on src/Storages/MergeTree/ReplicatedMergeTreeSink.cpp:
```diff
@@ -952,32 +923,47 @@ std::pair<std::vector<String>, bool> ReplicatedMergeTreeSinkImpl<async_insert>::
+                        "Unknown status of part {} (Reason: {}). Data was written locally but we don't know the status in keeper. It will be verified in ~{} seconds.",
+                        part->name, multi_code, MAX_AGE_OF_LOCAL_PART_THAT_WASNT_ADDED_TO_ZOOKEEPER);
+            });
+            new_retry_controller.requestUnconditionalRetry();
```
Probably comment why calling `requestUnconditionalRetry ()` is necessary? afaiu, to guarantee that `actionAfterLastFailedRetry` is called in case no retries left

devcrafter reviewed (approved):
LGTM with comments

Algunenano commented on src/Common/ZooKeeper/ZooKeeper.h:
```diff
@@ -194,6 +195,8 @@ struct MultiReadResponses
 /// Methods with names not starting at try- raise KeeperException on any error.
 class ZooKeeper
 {
+    friend class DB::ZooKeeperWithFaultInjection;
```
:+1:

Algunenano committed:
Review improvements

Algunenano committed:
Merge remote-tracking branch 'blessed/master' into backup_1

Algunenano commented on src/Backups/RestoreCoordinationRemote.cpp:
```diff
@@ -43,11 +44,9 @@ RestoreCoordinationRemote::RestoreCoordinationRemote(
             if (my_is_internal)
             {
                 String alive_node_path = my_zookeeper_path + "/stage/alive|" + my_current_host;
+                zk->deleteEphemeralNodeIfContentMatches(alive_node_path, "");
```
I would rather not hide it for now. In some places the ephemeral node is created as part of a transaction and in other places we might not want to automatically recreate the ephemeral node. It also introduces extra round trips that are not necessary if you are not recovering from an error (one to check, one to remove or waiting for a long time for it to be removed).

I'll refactor it a bit

Algunenano commented:
Working on rebasing the PR after the fix for zero-copy locks

Algunenano committed:
Merge remote-tracking branch 'blessed/master' into backup_1

Algunenano committed:
Adapt from HEAD

Algunenano committed:
Fix CLICKHOUSE_KEEPER_CLIENT with CLICKHOUSE_BINARY

Algunenano committed:
WIP: Remove UNCERTAIN_COMMIT in INSERT

Algunenano committed:
Partially revert "make stages commit"

This reverts commit e4becc01ba97c1a093810ac2d278b93c9c426e04.

Algunenano committed:
Implement retries when ZK connection fails without committing the transaction

Algunenano commented:
Reimplemented ZK retries to `tryMulti` both for before and after failures on top of @CheSema 's recent changes. So far so good. I had to recover `rollbackPartsToTemporaryState`

Algunenano committed:
Change check order in replication.lib

If there are no tables left merge() will keep throwing errors constantly

Algunenano committed:
Fix error on retries due to TABLE_IS_READ_ONLY

Algunenano commented:
Everything green except `Integration tests (asan) [4/4]` which failed because of `test_postgresql_replica_database_engine_1/test.py::test_many_concurrent_queries` which I don't see how it can be related.

Marking it as ready again after the last changes. cc @devcrafter in case you want to review the latest changes before merging

Algunenano commented:
Need to rebase on top of [57764](https://github.com/ClickHouse/ClickHouse/pull/57764) which fixes some of the same problems found and fixed in this PR

Algunenano committed:
Merge remote-tracking branch 'blessed/master' into backup_1

tavplubix commented:
> test_postgresql_replica_database_engine_1/test.py::test_many_concurrent_queries which I don't see how it can be related.

Even if it's not related, let's not ignore it. We can start with using [CI DB](https://play.clickhouse.com/play?user=play#CldJVEgKICAgICd0ZXN0X3Bvc3RncmVzcWxfcmVwbGljYV9kYXRhYmFzZV9lbmdpbmVfMScgQVMgbmFtZV9zdWJzdHIsCiAgICA5MCBBUyBpbnRlcnZhbF9kYXlzLAogICAgKCdTdGF0ZWxlc3MgdGVzdHMgKGFzYW4pJywgJ1N0YXRlbGVzcyB0ZXN0cyAoYWRkcmVzcyknLCAnU3RhdGVsZXNzIHRlc3RzIChhZGRyZXNzLCBhY3Rpb25zKScpIEFTIGJhY2twb3J0X2FuZF9yZWxlYXNlX3NwZWNpZmljX2NoZWNrcwpTRUxFQ1QKICAgIHRvU3RhcnRPZkRheShjaGVja19zdGFydF90aW1lKSBBUyBkLAogICAgY291bnQoKSwKICAgIGdyb3VwVW5pcUFycmF5KHB1bGxfcmVxdWVzdF9udW1iZXIpIEFTIHBycywKICAgIGFueShyZXBvcnRfdXJsKQpGUk9NIGNoZWNrcwpXSEVSRSAoKG5vdygpIC0gdG9JbnRlcnZhbERheShpbnRlcnZhbF9kYXlzKSkgPD0gY2hlY2tfc3RhcnRfdGltZSkgQU5EIChwdWxsX3JlcXVlc3RfbnVtYmVyIE5PVCBJTiAoCiAgICBTRUxFQ1QgcHVsbF9yZXF1ZXN0X251bWJlciBBUyBwcm4KICAgIEZST00gY2hlY2tzCiAgICBXSEVSRSAocHJuICE9IDApIEFORCAoKG5vdygpIC0gdG9JbnRlcnZhbERheShpbnRlcnZhbF9kYXlzKSkgPD0gY2hlY2tfc3RhcnRfdGltZSkgQU5EIChjaGVja19uYW1lIElOIChiYWNrcG9ydF9hbmRfcmVsZWFzZV9zcGVjaWZpY19jaGVja3MpKQopKSBBTkQgKHBvc2l0aW9uKHRlc3RfbmFtZSwgbmFtZV9zdWJzdHIpID4gMCkgQU5EICh0ZXN0X3N0YXR1cyBJTiAoJ0ZBSUwnLCAnRVJST1InLCAnRkxBS1knKSkKR1JPVVAgQlkgZApPUkRFUiBCWSBkIERFU0MK) to find when it started to fail. And it helps to find the reason quite easily: https://github.com/ClickHouse/ClickHouse/pull/57568#issuecomment-1854518149

Algunenano commented:
The new failures appeared with the merge with master and they look related. Need to investigate more, but I took master's way of improving ZK retries on backup coordination instead of the one in this PR so most likely the problem is there.

Algunenano committed:
Merge remote-tracking branch 'blessed/master' into backup_1

Algunenano committed:
Leave backups as HEAD

Algunenano committed:
Merge remote-tracking branch 'blessed/master' into backup_1

Algunenano commented:
Failures:
* `Integration tests (asan) [3/4]` -> test_storage_hdfs/test.py::test_write_table. Fixed in https://github.com/ClickHouse/ClickHouse/pull/55370
* `Integration tests (asan) [4/4] ` -> `test_storage_azure_blob_storage/test_cluster.py::test_unset_skip_unavailable_shards` -> Fixed in https://github.com/ClickHouse/ClickHouse/pull/57895
* `Integration tests (release) [4/4]` -> `test_storage_azure_blob_storage/test_cluster.py::test_unset_skip_unavailable_shards` -> Fixed in https://github.com/ClickHouse/ClickHouse/pull/57895
* `Stateless tests (aarch64) ` -> `00002_log_and_exception_messages_formatting`. I'll have a look in a different PR as it's not related to the changes.

I would love to not have to rebase this again :wink:

devcrafter commented:
> I would love to not have to rebase this again 😉

It'd be nice to understand what has changed from latest review

Algunenano commented:
> It'd be nice to understand what has changed from latest review

The last 2 rebases:

* https://github.com/ClickHouse/ClickHouse/pull/56917/commits/9d8d5dfdf63a526510fe4915f85897c68f98df3a + https://github.com/ClickHouse/ClickHouse/pull/56917/commits/923c3b7b32ea70d77e4e54ade201b2464e0f334d -> Reimplemented retries in RMTSink (commit) when the connection to ZK fails on tryMulti(). This had been already implemented but it was discovered that old retries weren't handling zero copy nodes correctly and the function was refactored completely, so retries had to be introduced again.

* Last one -> https://github.com/ClickHouse/ClickHouse/pull/56917/commits/3200933144fb3ab4e0d8ec5ce78d0ff9e5d528a6 -> Removes any changes vs master under `src/Backups` as the ones I had introduced have been also introduced in different PRs in the meantime.

devcrafter commented on src/Storages/MergeTree/ReplicatedMergeTreeSink.cpp:
```diff
@@ -914,14 +892,60 @@ std::pair<std::vector<String>, bool> ReplicatedMergeTreeSinkImpl<async_insert>::
+            });
+
+            /// Independently of how many retries we had left we want to do at least one check of this inner retry
+            /// at least once so a) we try to verify at least once if metadata was written and b) we set the proper
```
It's too many "at least once"

devcrafter commented on src/Storages/MergeTree/ReplicatedMergeTreeSink.cpp:
```diff
@@ -1019,22 +1038,6 @@ std::pair<std::vector<String>, bool> ReplicatedMergeTreeSinkImpl<async_insert>::
     {
         zookeeper->setKeeper(storage.getZooKeeper());
```
I'm not sure about moving this block to `lock and commit` stage. Probably, you could explain

Algunenano replied:
I moved it because the table being in read only only matters when we are going to add a new part/commit to ZK. If we only have checks left (for example in `resolve_duplicate_stage`) it doesn't make sense to not continue verifying because the table is stopped, as we can still return a proper response to the client.

A similar thing will happen for quorum for example (but it's not in this PR to keep in more concise).  It's ok to check for quorum if the write succeeded but the table was set in read only after that.

devcrafter reviewed (approved):
@Algunenano My comments are mostly minor, except probably https://github.com/ClickHouse/ClickHouse/pull/56917#discussion_r1428261563. But if you think it's fine, please feel free to fix comments in follow-up PR

Algunenano commented on src/Storages/MergeTree/MergeTreeData.cpp:
```diff
@@ -6339,6 +6358,24 @@ MergeTreeData::Transaction::Transaction(MergeTreeData & data_, MergeTreeTransact
+{
+    if (!isEmpty())
+    {
+        WriteBufferFromOwnString buf;
```
I recovered the code by reverting partially the commit that removed it and kept it as-is for simplicity, but it could be improved / iterated in the future.

Algunenano committed:
Review improvements

Algunenano commented:
Failures:
* AST fuzzer (ubsan) -> https://github.com/ClickHouse/ClickHouse/issues/57705
* Stateless tests (debug) [3/5]
  * `01508_partition_pruning_long`. It seems the queries just took too long and the test reached the 10m limit before finishing them all.
  * `01414_mutations_and_errors_zookeeper`: The mutation was cancelled internally (due to large sleep) before it was killed externally, so the log message is missing.
* `Upgrade check (debug)`: I see a OOM Kill in stress.log and I see a logical error in gdb.log but I can't see if they are related or what binary and data was being run. This test is really hard to debug/understand in general (https://github.com/ClickHouse/ClickHouse/issues/57981)
